### ISSUE TITLE
🚀 Gathering Feedback - Cloning Entries

### ISSUE THREAD

**Issue by ops-andy** [ExpressionEngine/ExpressionEngine#1077]
There have been many requests to add a Clone Entries feature to ExpressionEngine. Unfortunately, adding the ability to do this and have it work with many 3rd party add-ons is not a simple solution. However, due to the popularity of this request, the ExpressionEngine team really wants to get feedback from the community as to use cases and how you see a Clone Entries feature being implemented.

Thanks!

**Comment by pyrobob83** [comment]
IMO, the ideal scenario would literally to be able to clone any entry including all settings and all fields (native and third party) and either automatically incrementing/appending the Title & URL Title (eg start_entry_url_title_clone or similar) or requiring those as parameters as part of the process.  Of course ideally addons would need to handle this too, so if you were cloning a Structure 'page', the new page is added to the Structure tree with the same parent and other Structure-specific settings.

The practical application of this in my experience is clients wanting to mirror an existing page that we built out for them as part of the initial site build where they like the exact layout and want to copy the layout for a new page.  So cloning, then going back through and updating content.  These days most of our content is in Fluid Fields, with pretty flexible layouts and various settings within the individual fields that affect layout, so rebuilding an entire page from scratch can be daunting for some editors.

**Comment by ops-andy** [comment]
@pyrobob83 this seems to be the common use case, especially around Fluid Fields.  What are your thoughts around having Entry Templates, where you could lay out fluid fields, grid rows, etc the way they should be and the user would just fill in the content. That way they wouldn't have to clear out all the fields, just fill in the blanks.

**Comment by pyrobob83** [comment]
@ops-andy I'm not sure I'm understanding what you're talking about - are you saying Entry Templates would let devs pre-define specific layouts?  If so, I'm not sure that'd be particularly useful in our case - we try to let our layouts be flexible, not rigid.  If those pre-defined layouts practically just mean specific individual fields get added to the Fluid for population, I don't think that'd help either as our process (like many others I've come cross) is to have options within individual fields available to the Fluid that affect layout (where the flexibility comes from) - eg:

![image](https://user-images.githubusercontent.com/8193519/118310004-40881700-b4ab-11eb-829c-fe694253a29d.png)

So simply adding the 'Single Content Block' to the 'Page Builder' wouldn't help.  Plus I'm imagining those 'Entry Templates' have to be pre-defined by the devs rather than allowing the editor to duplicate something they've created elsewhere using the various options we as devs made available to them.  Does that make sense?

**Comment by ops-andy** [comment]
That's a great perspective on this. What if the users could easily define their own templates? For example, if they could choose to make a template out of an entry they created.

**Comment by murtaugh** [comment]
I'd say maybe every other client requests the ability to clone entries. My longest-term client is a church, and — as you might guess — they hold very similar events on a weekly basis. Similar but not identical, so cloning those entries would be very useful.

Another scenario is for product inventory; lots of products with similar, but not identical, attributes.

**Comment by murtaugh** [comment]
Another example of a client who needed cloning was a restaurant that published a new menu every week.

**Comment by murtaugh** [comment]
This seems like a tangent, but I would also love to be able to set up fluid fields with specific fields in place for new entries. I wouldn't require these fields, the user could still use the fluid field as normal, but being able to present fields that are going to be used 90% of the time would be very useful. (Or maybe a fluid field button that reveals a set of fields, instead of just one?)

**Comment by pyrobob83** [comment]
@ops-andy By 'template' are you referring to basically adding a pre-defined group of blank individual fields into a Fluid field?  If so, that could help in some scenarios and be a good starting point for editors, but a) it still doesn't help with the layout-affecting options inside those fields (unless a 'template' allows you to set default values?) and b) it means a lot of extra work, either for the dev setting those up or the editor creating their own.  If I'm understanding these 'templates' correctly, I could see them being helpful in limited scenarios, but cloning would be so much more useful across the board.  Many of the things clients have wanted to be able to clone weren't in Fluid Fields at all - like things Tim just mentioned above.

**Comment by ops-andy** [comment]
> @ops-andy By 'template' are you referring to basically adding a pre-defined group of blank individual fields into a Fluid field?

It's all up for discussion :) . But yes, I'm thinking you could lay out an entry with fluid field, grids, anything as populated or structured as much as you want. So if you want to allow for flexibility, but want to remind the user of the ideal layout for the design they signed off on, then your template might already have that structured out (with or without content in those blocks).

> This seems like a tangent, but I would also love to be able to set up fluid fields with specific fields in place for new entries. I wouldn't require these fields, the user could still use the fluid field as normal, but being able to present fields that are going to be used 90% of the time would be very useful. (Or maybe a fluid field button that reveals a set of fields, instead of just one?)

this is more the use case for templates that we've been thinking about. Except not fluid field specific, rather it would a template of the whole entry, but you could lay out fluid fields in an ideal way.

this is good stuff and continues to give good perspective. We want to provide the best solution. While everyone immediately reaches for cloning, we want to make sure there aren't other problems that need to be solved or ways to make it easier for the end-user.

**Comment by pyrobob83** [comment]
@ops-andy If a 'template' could let you define everything that shows in a 'new entry' publish form including which fields are added to Fluids **and** include default content / field selections like images in file fields, the number of rows a grid starts with, dropdown selections, toggle position, etc - both for standalone entry fields and Fluid sub-fields - then I could see that being useful too, but more for what you mentioned above about reminding the editor of the approved layouts.  I still feel that's very different and would be used a lot less than if a true 'clone entry' function was available.

**Comment by strudelmedia** [comment]
For me, the template wouldn't help (although I can imagine its use!). What @murtaugh was mentioning at the top is what I would need, which I think would be more simple anyway. One of my main use cases is for an online photography school; we often re-run a class or need to quickly duplicate an  entry to provide a special price to someone. So for this, cloning the entry with all the extant info is important; we just go in and change the dates/price/classcode whatever. Without wCloner function, the workflow would be waaaaay worse. So we need to wait on entry cloning before upgrading from ee5.

**Comment by blue-dreamer** [comment]
My use case is for ecommerce sites where authors regularly add "new" products entries which have similar content but only a few data points change such as prices and/or categories/classifications. It's a lot quicker to clone an existing entry and tweak a few fields rather than have to add all info for every new entry every time.

Ideally for me cloning would duplicate every field value, including categories, entry dates, statuses etc.

**Comment by mattrandles** [comment]
I like the idea of entry templates a lot for repetetive posts and giving editors more autonomy over new pages but I'm often asked how to clone an entry. On cloning, I think it would be nice to have the ability to clone from the channel entries screen: either auto-appending url titles and adding it to the list, or sending the user to a pre populated "new entry" (akin to how creating a new template can either be saved, or saved and edited).

The main use case for cloning on my client projects seems to be creating short-live staging pages where many people will give input on updates before they replace the live page. They'll create a "-v2" entry set to closed, pass on a preview URL (just a copy of entry template with closed entries enabled) and get sign off before renaming items to make the new entry the 'live' page. Some clients use the publisher add-on to avoid this, but some don't want to give out any sort of admin login to preview an entry.

**Comment by ic360-com** [comment]
At the moment, I'm using the wCloner add-on by Willow Light Studios to clone Entries, which is pretty much essential for the all the ecommerce sites I build. I'm not sure it's EE6 compatible yet though. It works very well, but doesn't clone the images from the Channel Images fields.

My gut feeling, is if you can get a Cloning function working with all the native EE fields first, I imagine you might be able to expand this with time, to get it working with other Add-Ons (especially the EE Harbour ones).
Or, would it be possible to create some kind of Hook (not sure if this is the correct terminology) that the 3rd party Add-On creators could use, to enable their add-ons' fields to also be Cloned?

In terms of simplistic pseudo code, I'm imaging it could all be done like this:
Initiate cloning function -> Copy all Entry field data into a buffer array/table -> Copy all associated images/files into a temp folder -> write all the data to new Entry -> delete all temp files.
But I'm no coder, so I might be way off!  ;-)

**Comment by jkoivisto** [comment]
One of the travel sites we have uses wcloner on their ee5 site daily. They said it was that ability that cut their workload by 50%. They set up tours and such for different groups, and once they set up the first tour (many per destination, several destinations per program, etc.), they clone the entry, update a few group/team relationships, change specific group description content, and move on to the next tour for that program.

**Comment by jkoivisto** [comment]
Another site that I manage with wCloner uses it mostly for newsletters. The entry from the previous month is cloned, past items are deleted, new items added, and the date updated. Being able to start with an entry already including all the previous month's data saves them some think time and proofreading time.

**Comment by litzinger** [comment]
Creating a template for an entry and cloning an entry to me are 2 very different things. By definition a clone should be a 100% copy of an existing entry, and wCloner does that pretty well by basically saving the existing entry as a new entry. Whatever native cloning ends up looking like, I hope it has hooks so 3rd party devs can tap into it if needed.

The revision system can be leveraged to create a clone. When you're viewing a revision it's loading the previous state of a saved entry. So injecting a revision into a new entry to create a clone should be pretty straight forward. It would just have to ensure that  the new data being inserted had unique IDs.

**Comment by PDXMan1** [comment]
Our most common use case is when a client is reorganizing the content structure of an existing site and needs to move a dozen or so entries around to different channels (with different categories). Yesterday a client had an entry with a grid field containing rows that include a name, photo/image, and bio of 20 volunteers, and our client didn't want to have to recreate all the content in those 20 grid rows in a new channel. Also events, like when a jazz group plays at the same club each week/month but uses different players in the band each time.  Would love to see this become a core feature of EE.

**Comment by frshjb373** [comment]
Nothing really new here, but a plus one for ability to clone entries. We're working on an EE theme builder/boilerplate that has specific blocks (using bloqs) and we would like to offer the ability for user to quickly clone pages (saving page templates to a library would also be nice feature as well as @litzinger distinguished above), but what I'm talking about is an exact copy of an existing entry.

**Comment by creativearc-paul** [comment]
We have a client that stores an entry, with a single grid, of fabric patterns. But that grid can easily have upwards of 100 rows. So in EE5, that's a huge bonus that they can use wcloner. This is likely a showstopper from them moving to EE6. I think cloning that only supported 1st party fieldtypes would alone be worth it.
